File: PadraoGof/Criacional/Builder/builder.js
/* 
--------------------------------------------------------------------------------------------------
Fatec Diadema - Luigi Papaiz 
Curso : Desenvolvimento de Software Multiplataforma - DSM 
Disciplina: Técnica de Programação II
Nome: Kaiki Kenji       
Data: 08/03/2024
Descrição: Aplicação do padrão GoF - Criacional - Builder desenvolvido como exemplo na aula.
Enunciado: No programa abaixo, é criado um software que simula a construção
de um carro, onde o builder nos permitirá definir diferentes partes do
carro de forma flexível. Começa definindo as diferentes partes do carro e
na sequência se cria o builder, que permite construir diferentes tipos de
carros; Se cria uma classe Carro que receberá as partes construidas pelo
builder e para fechar, usa-se o Builder para construir diferentes tipos de carros
--------------------------------------------------------------------------------------------------
*/

/* Define as partes de um carro */
class Motor {
    constructor(tipo) {
        this.tipo = tipo;
    }
}

class Carroceria {
    constructor(estilo) {
        this.estilo = estilo;
    }
}

class Rodas {
    constructor(tamanho) {
        this.tamanho = tamanho;
    }
}

class CorVeiculo {
    constructor(cor) {
        this.cor = cor;
    }
}

class Nome {
    constructor(nome) {
        this.nome = nome
    }
}

/* Builder */
class CarroBuilder {
    constructor() {
        this.motor = null;
        this.carroceria = null;
        this.rodas = null;
        this.corVeiculo = null;
        this.nome = null;
    }

    adicionarMotor(tipo) {
        this.motor = new Motor(tipo);
        return this;
    }
    adicionarCarroceria(estilo) {
        this.carroceria = new Carroceria(estilo);
        return this;
    }
    adicionarRodas(tamanho) {
        this.rodas = new Rodas(tamanho);
        return this;
    }
    adicionarCorVeiculo(cor) {
        this.corVeiculo = new CorVeiculo(cor);
        return this;
    }
    adicionarNome(nome) {
        this.nome = new Nome(nome);
        return this;
    }

    montarCarro() {
        return new Carro(this.motor, this.carroceria, this.rodas, this.corVeiculo, this.nome)
    }
}

/* Construindo o Carro */
class Carro {
    constructor(motor, carroceria, rodas, corVeiculo, nome) {
        this.motor = motor;
        this.carroceria = carroceria;
        this.rodas = rodas;
        this.corVeiculo = corVeiculo;
        this.nome = nome;
    }

    montarDetalhes() {
        console.log(`
            Carro com Motor: ${this.motor.tipo}, 
            Carroceria: ${this.carroceria.estilo}
            Rodas de Tamanho: ${this.rodas.tamanho}
            Cor: ${this.corVeiculo.cor} 
            Nome: ${this.nome.nome}
            \n`
        )
    }
}

/* Interface do usuario - Utilização do Padrão:*/
const builder = new CarroBuilder();

/* Construção do Carro Esportivo */
const carroPadrao = builder
    .adicionarMotor('-------')
    .adicionarCarroceria('-------')
    .adicionarRodas('-------')
    .adicionarCorVeiculo('-------')
    .adicionarNome('-------')
    .montarCarro();

const carroEsportivo = builder
    .adicionarMotor('V8')
    .adicionarCarroceria('Esportiva')
    .adicionarRodas(18)
    .adicionarCorVeiculo('Preto')
    .adicionarNome('Esportivo')
    .montarCarro();

/* Construção do Carro Sedan */
const carroPadrao2 = builder
    .adicionarMotor('-------')
    .adicionarCarroceria('-------')
    .adicionarRodas('-------')
    .adicionarCorVeiculo('-------')
    .adicionarNome('-------')
    .montarCarro();

const carroSedan = builder
    .adicionarMotor('4 Cilindros')

    .adicionarNome('Sedan')
    .montarCarro();

/* Mostrar os Veiculos Construidos */
carroPadrao.montarDetalhes();
carroEsportivo.montarDetalhes();
carroPadrao2.montarDetalhes();
carroSedan.montarDetalhes();

File: PadraoGof/Criacional/Builder/builder.py
#--------------------------------------------------------------------------------------------------
#Fatec Diadema - Luigi Papaiz 
#Curso : Desenvolvimento de Software Multiplataforma - DSM 
#Disciplina: Técnica de Programação II
#Nome: Kaiki Kenji       
#Data: 08/03/2024
#Descrição: Aplicação do padrão GoF - Criacional - Builder desenvolvido como exemplo na aula.
#Enunciado: No programa abaixo, é criado um software que simula a construção
#de um carro, onde o builder nos permitirá definir diferentes partes do
#carro de forma flexível. Começa definindo as diferentes partes do carro e
#na sequência se cria o builder, que permite construir diferentes tipos de
#carros; Se cria uma classe Carro que receberá as partes construidas pelo
#builder e para pra fechar, usa-se o Builder para construir diferentes tipos de carros
#--------------------------------------------------------------------------------------------------

# Define as partes de um carro 
class NomeCarro:
    def __init__(self,nome):
        self.nome = nome

class Motor:
    def __init__(self, tipo):
        self.tipo = tipo

class Carroceria:
    def __init__(self,estilo):
        self.estilo = estilo

class Rodas:
    def __init__(self,tamanho):
        self.tamanho = tamanho

class CorVeiculo:
    def __init__(self,cor):
        self.cor = cor

# Builder 
class Carro:
    def __init__(self, nome, motor, carroceria, rodas, corVeiculo):
        self.nome = nome
        self.motor = motor
        self.carroceria = carroceria
        self.rodas = rodas
        self.corVeiculo = corVeiculo 

    def mostrarDetalhes(self):
        print(
            f"Nome: {self.nome.nome}\n" + 
            f"Motor: {self.motor.tipo}\n"+
            f"Carroceria: {self.carroceria.estilo}\n" +
            f"Rodas: {self.rodas.tamanho}\n"+
            f"Cor: {self.corVeiculo.cor}\n"
            )

class CarroBuilder:
    def __init__(self):
        self.nome = None
        self.motor = None
        self.carroceria = None
        self.rodas = None
        self.corVeiculo = None
    
    def adicionarNome(self,nome):
        self.nome = NomeCarro(nome)
        return self
    
    def adicionarMotor(self,tipo):
        self.motor = Motor(tipo)
        return self
    
    def adicionarCarroceria(self,estilo):
        self.carroceria = Carroceria(estilo)
        return self
    
    def adicionarRodas(self,tamanho):
        self.rodas = Rodas(tamanho)
        return self
    
    def adicionarCorVeiculo(self, cor):
        self.corVeiculo = CorVeiculo(cor)
        return self
    
    def montarCarro(self):
        return Carro(self.nome, self.motor, self.carroceria, self.rodas, self.corVeiculo)

# Construindo o Carro 

# Interface do usuario - Utilização do Padrão: 
build = CarroBuilder()

# Construção do Carro Esportivo 
carroEsportivo = build\
    .adicionarNome('Esportivo')\
    .adicionarMotor('V8')\
    .adicionarCarroceria('Esportivo')\
    .adicionarRodas(18)\
    .adicionarCorVeiculo("Preto")\
    .montarCarro()

# Construção do Carro Sedan 
carroSedan = build\
    .adicionarNome('Sedan')\
    .adicionarMotor('Motor 4 cilindro')\
    .adicionarCarroceria('Sedan')\
    .adicionarRodas(16)\
    .adicionarCorVeiculo("bege")\
    .montarCarro()

# Mostrar os Veiculos Construidos 

carroEsportivo.mostrarDetalhes()
carroSedan.mostrarDetalhes()